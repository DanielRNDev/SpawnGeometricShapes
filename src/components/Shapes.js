import React, { Component } from 'react'
import {
  View,
  PanResponder,
  StyleSheet,
} from 'react-native'
import Svg, {
  Circle,
  Rect,
  Image,
  Defs,
  ClipPath,
  Polygon
} from 'react-native-svg'
import * as Animatable from 'react-native-animatable'
import { SQUARE, CIRCLE, TRIANGLE } from '../constants'

const styles = StyleSheet.create({
  shapeWrapper: {
    position: 'absolute'
  }
})
const AnimatableView = Animatable.createAnimatableComponent(View)

export default class Shapes extends Component {
  constructor(props) {
    super(props)
    const { positionX, positionY } = this.props

    this.state = {
      left: positionX,
      top: positionY,
      pressed: false,
    }

    this.previousLeft = positionX
    this.previousTop = positionY
  }

  handleStartShouldSetPanResponder = (event) => {
    event && event.preventDefault && event.preventDefault()
    return true
  }

  handleMoveShouldSetPanResponder = (event) => {
    event && event.preventDefault && event.preventDefault()
    return true
  }

  handlePanResponderGrant = () => {
    this.setState({
      pressed: true,
    })
  }

  handlePanResponderMove = (event, gestureState) => {
    this.setState({
      left: this.previousLeft + gestureState.dx,
      top: this.previousTop + gestureState.dy,
    })
  }

  handlePanResponderEnd = (event, gestureState) => {
    const { updateShapeValue } = this.props

    this.setState({
      pressed: false,
    })
    this.previousLeft += gestureState.dx
    this.previousTop += gestureState.dy

    updateShapeValue({
      positionX: this.previousLeft,
      positionY: this.previousTop
    })
  }

  panResponder = PanResponder.create({
    onStartShouldSetPanResponder: this.handleStartShouldSetPanResponder,
    onMoveShouldSetPanResponder: this.handleMoveShouldSetPanResponder,
    onPanResponderGrant: this.handlePanResponderGrant,
    onPanResponderMove: this.handlePanResponderMove,
    onPanResponderRelease: this.handlePanResponderEnd,
    onPanResponderTerminate: this.handlePanResponderEnd,
  })

  renderSquare = () => {
    const { image, size, color } = this.props

    return (
      <Svg width={size} height={size}>
        {
          image
            ? <Image
              width={size}
              height={size}
              preserveAspectRatio="xMidYMid slice"
              href={{ uri: image }}
            />
            : <Rect
              width={size}
              height={size}
              fill={color}
            />
        }
      </Svg>
    )
  }

  renderCircle = () => {
    const { size, color } = this.props
    const r = size / 2

    return (
      <Svg height={size} width={size}>
        <Circle cx={r} cy={r} r={r} fill={color} />
      </Svg>
    )
  }

  renderTriangle = () => {
    const { image, size, color, points } = this.props

    if (image) {
      return (
        <Svg height={size} width={size}>
          <Defs>
            <ClipPath id="clip">
              <Polygon points={points} />
            </ClipPath>
          </Defs>
          <Image
            width={size}
            height={size}
            preserveAspectRatio="xMidYMid slice"
            href={{ uri: image }}
            clipPath="url(#clip)"
          />
        </Svg>
      )
    }

    return (
      <Svg height={size} width={size}>
        <Polygon points={points} fill={color} />
      </Svg>
    )
  }

  renderShape = () => {
    const { type } = this.props

    switch (type) {
      case SQUARE:
        return this.renderSquare()

      case CIRCLE:
        return this.renderCircle()

      case TRIANGLE:
        return this.renderTriangle()

      default:
        return null
    }
  }

  render() {
    return (
      <AnimatableView
        animation="bounceIn"
        {...this.panResponder.panHandlers}
        style={[
          styles.shapeWrapper,
          {
            left: this.state.left,
            top: this.state.top,
          }
        ]}
      >
        {this.renderShape()}
      </AnimatableView>
    )
  }
}
