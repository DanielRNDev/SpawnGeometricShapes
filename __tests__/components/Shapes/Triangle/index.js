import 'react-native';
import React from 'react'
import Shapes from '../../../../src/components/Shapes'
import renderer from 'react-test-renderer'

describe('Shapes Components', () => {
  it('renders triangle with image correctly', () => {
    const data = {
      type: 'TRIANGLE',
      size: 150,
      positionX: 50,
      positionY: 100,
      color: `#FFFFFF`,
      image: 'https://reactnative.dev/img/tiny_logo.png',
    }
    const tree = renderer.create(<Shapes {...data} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
