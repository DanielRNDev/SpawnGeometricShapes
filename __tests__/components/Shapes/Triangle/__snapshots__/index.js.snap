// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shapes Components renders triangle with image correctly 1`] = `
<View
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  style={
    Object {
      "left": 50,
      "opacity": 0,
      "position": "absolute",
      "top": 100,
      "transform": Array [
        Object {
          "scale": 0.3,
        },
      ],
    }
  }
>
  <RNSVGSvgView
    bbHeight={150}
    bbWidth={150}
    focusable={false}
    height={150}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 150,
          "width": 150,
        },
      ]
    }
    width={150}
  >
    <RNSVGGroup>
      <RNSVGDefs>
        <RNSVGClipPath
          name="clip"
        >
          <RNSVGPath
            d=""
          />
        </RNSVGClipPath>
      </RNSVGDefs>
      <RNSVGImage
        align="xMidYMid"
        clipPath="clip"
        height={150}
        meetOrSlice={1}
        src={
          Object {
            "uri": "https://reactnative.dev/img/tiny_logo.png",
          }
        }
        width={150}
        x={0}
        y={0}
      />
    </RNSVGGroup>
  </RNSVGSvgView>
</View>
`;
